app/Models/Extensions/SizeVariants.php

```diff
@@ -0,0 +1,243 @@
+<?php
+
+namespace App\Models\Extensions;
+
+use App\Models\Photo;
+use App\Models\SizeVariant;
+use Illuminate\Contracts\Support\Arrayable;
+use Illuminate\Database\Eloquent\Collection;
+use JsonSerializable;
+
+/**
+ * Class SizeVariants.
+ */
+class SizeVariants implements Arrayable, JsonSerializable
+{
+	/** @var Photo the parent object this object is tied to */
+	private Photo $photo;
+
+	private ?SizeVariant $original = null;
+	private ?SizeVariant $medium2x = null;
+	private ?SizeVariant $medium = null;
+	private ?SizeVariant $small2x = null;
+	private ?SizeVariant $small = null;
+	private ?SizeVariant $thumb2x = null;
+	private ?SizeVariant $thumb = null;
+
+	/**
+	 * SizeVariants constructor.
+	 *
+	 * @param Photo                        $photo        the parent object
+	 *                                                   this object is tied to
+	 * @param Collection<SizeVariant>|null $sizeVariants a collection of size
+	 *                                                   variants
+	 */
+	public function __construct(Photo $photo, ?Collection $sizeVariants = null)
+	{
+		$this->photo = $photo;
+		if ($sizeVariants) {
+			/** @var SizeVariant $sizeVariant */
+			foreach ($sizeVariants as $sizeVariant) {
+				$this->add($sizeVariant);
+			}
+		}
+	}
+
+	public function add(SizeVariant $sizeVariant): void
+	{
+		if ($sizeVariant->photo_id !== $this->photo->id) {
+			throw new \UnexpectedValueException('ID of owning photo does not match');
+		}
+		$sizeVariant->setRelation('photo', $this->photo);
+
+		switch ($sizeVariant->size_variant) {
+			case SizeVariant::ORIGINAL:
+				$ref = &$this->original;
+				break;
+			case SizeVariant::MEDIUM2X:
+				$ref = &$this->medium2x;
+				break;
+			case SizeVariant::MEDIUM:
+				$ref = &$this->medium;
+				break;
+			case SizeVariant::SMALL2X:
+				$ref = &$this->small2x;
+				break;
+			case SizeVariant::SMALL:
+				$ref = &$this->small;
+				break;
+			case SizeVariant::THUMB2X:
+				$ref = &$this->thumb2x;
+				break;
+			case SizeVariant::THUMB:
+				$ref = &$this->thumb;
+				break;
+			default:
+				throw new \UnexpectedValueException('size variant ' . $sizeVariant . 'invalid');
+		}
+
+		if ($ref && $ref->id !== $sizeVariant->id) {
+			throw new \UnexpectedValueException('Another size variant of the same type has already been added');
+		}
+		$ref = $sizeVariant;
+	}
+
+	/**
+	 * Serializes this object into an array.
+	 *
+	 * @return array The serialized properties of this object
+	 */
+	public function toArray(): array
+	{
+		return [
+			'original' => $this->original ? $this->original->toArray() : null,
```

Comment:
Php8: https://stitcher.io/blog/php-8-nullsafe-operator :)
```suggestion
			'original' => $this->original?->toArray(),
```

Reply:
I know. But PHP 8 is not yet merged into master and currently the test suite still include PHP 7.4. I will iterate over all these, when PHP 8 is in master and master merged into this PR.

However, I wonder, if we really want to do that, because this will intentionally break any chance that Lychee runs on older versions than PHP 8.

Reply:
Well didn't we already established that the PR with php 8 is breaking 7.4 support anyway due to union types :') ?

Reply:
> Well didn't we already established that the PR with php 8 is breaking 7.4 support anyway due to union types :') ?

Yes, we did. But I am still not sure, if I like it. Anyway, after the PR for PHP 8 has been merged into master, I decide upon that, which fancy, new language feature of PHP 8 I will adopt.